rustowl/man/rustowl.1
.ie \n(.g .ds Aq \(aq
.el .ds Aq '
.TH "RustOwl Language Server" 1  "RustOwl Language Server 0.2.3" 
.SH NAME
RustOwl Language Server
.SH SYNOPSIS
\fBrustowl\fR [\fB\-\-stdio\fR] [\fB\-h\fR|\fB\-\-help\fR] [\fB\-V\fR|\fB\-\-version\fR] [\fIsubcommands\fR]
.SH DESCRIPTION
.SH OPTIONS
.TP
\fB\-\-stdio\fR

.TP
\fB\-h\fR, \fB\-\-help\fR
Print help
.TP
\fB\-V\fR, \fB\-\-version\fR
Print version
.SH SUBCOMMANDS
.TP
RustOwl Language Server\-check(1)
.TP
RustOwl Language Server\-clean(1)
.TP
RustOwl Language Server\-toolchain(1)
.TP
RustOwl Language Server\-completions(1)
Generate shell completions
.TP
RustOwl Language Server\-help(1)
Print this message or the help of the given subcommand(s)
.SH VERSION
v0.2.3
.SH AUTHORS
cordx56 <cordx56@cordx.cx>


CHANGELOG.md
<a name="unreleased"></a>
## [Unreleased]

### 🎨 Chores
- update changelog

### 🐞 Bug Fixes
- **reqwest:** dont depend on openssl-sys, use rustls for lower system deps

### 🚀 Features
- auto release changelogs, changelog generation
- shell completions and man pages


<a name="v0.2.3pre"></a>
## [v0.2.3pre] - 2025-04-24
### Reverts
- test workflow

### Pull Requests
- Merge pull request [#80](https://github.com/cordx56/rustowl/issues/80) from siketyan/ci/more-platform


<a name="v0.2.2"></a>
## [v0.2.2] - 2025-04-18

<a name="v0.2.2pre2"></a>
## [v0.2.2pre2] - 2025-04-18

<a name="v0.2.2pre"></a>
## [v0.2.2pre] - 2025-04-18
### ♻️ Code Refactoring
- streamline toolchain detection and correct cargo path

### 🚀 Features
- **toolchain:** add support for RUSTOWL_TOOLCHAIN_DIR to bypass rustup

### Pull Requests
- Merge pull request [#77](https://github.com/cordx56/rustowl/issues/77) from xBLACKICEx/flexible-toolchain


<a name="v0.2.1"></a>
## [v0.2.1] - 2025-04-15

<a name="v0.2.0"></a>
## [v0.2.0] - 2025-04-09
### ♻️ Code Refactoring
- add prefix to functions with commonly used names

### 🎨 Chores
- add require lsp
- remove calling `enable-rustowlsp-cursor`
- add `defgroup`
- add `provide`
- Migrate to Rust 2024

### 🐞 Bug Fixes
- package-requires

### Reverts
- messsage type
- neovim plugin function
- update install manual

### Pull Requests
- Merge pull request [#72](https://github.com/cordx56/rustowl/issues/72) from mawkler/neovim-version
- Merge pull request [#69](https://github.com/cordx56/rustowl/issues/69) from cordx56/feat/elim-rustup-call
- Merge pull request [#48](https://github.com/cordx56/rustowl/issues/48) from mawkler/lua-api
- Merge pull request [#62](https://github.com/cordx56/rustowl/issues/62) from Kyure-A/main
- Merge pull request [#61](https://github.com/cordx56/rustowl/issues/61) from AIDIGIT/nvim-hl-priorities
- Merge pull request [#60](https://github.com/cordx56/rustowl/issues/60) from AIDIGIT/main
- Merge pull request [#55](https://github.com/cordx56/rustowl/issues/55) from sorairolake/migrate-to-2024-edition


<a name="v0.1.4"></a>
## [v0.1.4] - 2025-02-22
### ♻️ Code Refactoring
- simplify HashMap insertion by using entry API

### Pull Requests
- Merge pull request [#54](https://github.com/cordx56/rustowl/issues/54) from uhobnil/main


<a name="v0.1.3"></a>
## [v0.1.3] - 2025-02-20
### 🎨 Chores
- remove duplicate code

### 🐞 Bug Fixes
- install the newest version

### Pull Requests
- Merge pull request [#53](https://github.com/cordx56/rustowl/issues/53) from uhobnil/main
- Merge pull request [#47](https://github.com/cordx56/rustowl/issues/47) from robin-thoene/fix/update-install-script


<a name="v0.1.2"></a>
## [v0.1.2] - 2025-02-19
### 🎨 Chores
- add the description for duplication
- add config.yaml
- add issue templae for feature requesting
- add labels to bug_report
- add issue templae for bug reporing

### 🐞 Bug Fixes
- s/enhancement/bug/
- update the introduction
- correct label
- remove redundant textarea
- update the information
- update the file extension
- s/rustowl/RustOwl/
- kill process when the client/server is dead

### Pull Requests
- Merge pull request [#35](https://github.com/cordx56/rustowl/issues/35) from chansuke/chore/add-issue-template
- Merge pull request [#42](https://github.com/cordx56/rustowl/issues/42) from uhobnil/main
- Merge pull request [#34](https://github.com/cordx56/rustowl/issues/34) from mtshiba/main
- Merge pull request [#26](https://github.com/cordx56/rustowl/issues/26) from Toyo-tez/main
- Merge pull request [#11](https://github.com/cordx56/rustowl/issues/11) from wx257osn2/clippy
- Merge pull request [#24](https://github.com/cordx56/rustowl/issues/24) from mawkler/main


<a name="v0.1.1"></a>
## [v0.1.1] - 2025-02-07

<a name="v0.1.0"></a>
## [v0.1.0] - 2025-02-05
### Pull Requests
- Merge pull request [#2](https://github.com/cordx56/rustowl/issues/2) from wx257osn2/support-windows


<a name="v0.0.5"></a>
## [v0.0.5] - 2025-02-02

<a name="v0.0.4"></a>
## [v0.0.4] - 2025-01-31

<a name="v0.0.3"></a>
## [v0.0.3] - 2025-01-30
### Pull Requests
- Merge pull request [#6](https://github.com/cordx56/rustowl/issues/6) from Jayllyz/build/enable-lto-codegen


<a name="v0.0.3pre"></a>
## [v0.0.3pre] - 2025-01-26
### Pull Requests
- Merge pull request [#5](https://github.com/cordx56/rustowl/issues/5) from mu001999-contrib/main


<a name="v0.0.2"></a>
## [v0.0.2] - 2025-01-23

<a name="v0.0.2pre"></a>
## [v0.0.2pre] - 2025-01-23

<a name="v0.0.1"></a>
## [v0.0.1] - 2024-11-13

<a name="vpre"></a>
## vpre - 2024-11-11

[Unreleased]: https://github.com/cordx56/rustowl/compare/v0.2.3pre...HEAD
[v0.2.3pre]: https://github.com/cordx56/rustowl/compare/v0.2.2...v0.2.3pre
[v0.2.2]: https://github.com/cordx56/rustowl/compare/v0.2.2pre2...v0.2.2
[v0.2.2pre2]: https://github.com/cordx56/rustowl/compare/v0.2.2pre...v0.2.2pre2
[v0.2.2pre]: https://github.com/cordx56/rustowl/compare/v0.2.1...v0.2.2pre
[v0.2.1]: https://github.com/cordx56/rustowl/compare/v0.2.0...v0.2.1
[v0.2.0]: https://github.com/cordx56/rustowl/compare/v0.1.4...v0.2.0
[v0.1.4]: https://github.com/cordx56/rustowl/compare/v0.1.3...v0.1.4
[v0.1.3]: https://github.com/cordx56/rustowl/compare/v0.1.2...v0.1.3
[v0.1.2]: https://github.com/cordx56/rustowl/compare/v0.1.1...v0.1.2
[v0.1.1]: https://github.com/cordx56/rustowl/compare/v0.1.0...v0.1.1
[v0.1.0]: https://github.com/cordx56/rustowl/compare/v0.0.5...v0.1.0
[v0.0.5]: https://github.com/cordx56/rustowl/compare/v0.0.4...v0.0.5
[v0.0.4]: https://github.com/cordx56/rustowl/compare/v0.0.3...v0.0.4
[v0.0.3]: https://github.com/cordx56/rustowl/compare/v0.0.3pre...v0.0.3
[v0.0.3pre]: https://github.com/cordx56/rustowl/compare/v0.0.2...v0.0.3pre
[v0.0.2]: https://github.com/cordx56/rustowl/compare/v0.0.2pre...v0.0.2
[v0.0.2pre]: https://github.com/cordx56/rustowl/compare/v0.0.1...v0.0.2pre
[v0.0.1]: https://github.com/cordx56/rustowl/compare/vpre...v0.0.1


lua/rustowl.lua
local lspconfig = require('lspconfig')
local util = require('lspconfig.util')
local configs = require('lspconfig.configs')

local hlns = vim.api.nvim_create_namespace('rustowl')
vim.api.nvim_set_hl(0, 'lifetime', { undercurl = true, sp = '#00cc00' })
vim.api.nvim_set_hl(0, 'imm_borrow', { undercurl = true, sp = '#0000cc' })
vim.api.nvim_set_hl(0, 'mut_borrow', { undercurl = true, sp = '#cc00cc' })
vim.api.nvim_set_hl(0, 'move', { undercurl = true, sp = '#cccc00' })
vim.api.nvim_set_hl(0, 'call', { undercurl = true, sp = '#cccc00' })
vim.api.nvim_set_hl(0, 'outlive', { undercurl = true, sp = '#cc0000' })

local hl_priorities = {
	lifetime = 200,
	imm_borrow = 201,
	mut_borrow = 202,
	move = 203,
	call = 204,
	outlive = 205,
}

local function show_rustowl(bufnr)
    bufnr = util.validate_bufnr(bufnr)
    local clients = util.get_lsp_clients { bufnr = bufnr, name = 'rustowl' }
    for _, client in ipairs(clients) do
        local line, col = unpack(vim.api.nvim_win_get_cursor(0))
        client.request(
            'rustowl/cursor',
            {
                position = {
                    line = line - 1,
                    character = col,
                },
                document = vim.lsp.util.make_text_document_params(),
            },
            function(err, result, ctx)
                if result ~= nil then
                    for _, deco in ipairs(result['decorations']) do
                        local start = { deco['range']['start']['line'], deco['range']['start']['character'] }
                        local finish = { deco['range']['end']['line'], deco['range']['end']['character'] }
                        vim.highlight.range(
                            bufnr,
                            hlns,
                            deco['type'],
                            start,
                            finish,
                            { regtype = "v", inclusive = true, priority = hl_priorities[deco["type"]] }
                        )
                    end
                end
            end,
            bufnr
        )
    end
end


local function rustowl_on_attach(hover, client, bufnr, idle_time_ms)
    local timer = nil
    local augroup = vim.api.nvim_create_augroup('RustOwlCmd', { clear = true })

    local function clear_timer()
        if timer then
            timer:stop()
            timer:close()
            timer = nil
        end
    end

    local function start_timer()
        clear_timer()
        timer = vim.uv.new_timer()
        timer:start(
            idle_time_ms,
            0,
            vim.schedule_wrap(
                function()
                    show_rustowl(bufnr)
                end
            )
        )
    end

    vim.api.nvim_create_autocmd(
        { 'CursorMoved', 'CursorMovedI' },
        {
            group = augroup,
            buffer = bufnr,
            callback = function()
                vim.api.nvim_buf_clear_namespace(bufnr, hlns, 0, -1)
                if hover == true then
                    start_timer()
                end
            end
        }
    )

    vim.api.nvim_create_autocmd(
        'BufUnload',
        {
            group = augroup,
            buffer = bufnr,
            callback = clear_timer
        }
    )

    start_timer()
end


if not configs.rustowl then
    configs.rustowl = {
        default_config = {
            cmd = { 'cargo', 'owlsp' },
            root_dir = lspconfig.util.root_pattern('Cargo.toml', '.git'),
            filetypes = { 'rust' },
            on_attach = function(client, bufnr)
            end,
        },
        idle_time = 2000,
    }

    local orig_setup = lspconfig.rustowl.setup
    lspconfig.rustowl.setup = function(user_opts)
        user_opts = user_opts or {}
        local user = user_opts.on_attach
        local trigger = user_opts.trigger or {
            hover = true,
        }
        local idle_time = tonumber(user_opts.idle_time) or 2000

        user_opts.on_attach = function(client, bufnr)
            rustowl_on_attach(trigger.hover, client, bufnr, idle_time)

            if type(user) == 'function' then
                user(client, bufnr)
            end
        end
        orig_setup(user_opts)
    end
end


return {
    rustowl_cursor = function(...)
        args = {...}
        bufnr = args[1] or vim.api.nvim_get_current_buf()
        show_rustowl(bufnr)
    end,
}
